cli/core/upgrade.go

```diff
@@ -94,7 +94,7 @@ func runUpgradeCommand(cmd *cobra.Command, args []string) {
 		}
 
 		_, err := core.PlatformUpgrade(context.Background(), r, output.ProgressBar(), output.TaskProgress())
-		if err == core.ErrAlreadyLatest {
+		if err.Message() == core.ErrAlreadyLatest.Error() {
```

Comment:
I'm not really happy with this, but I didn't come up with any better way. Any ideas?

Reply:
If we want to "tag" a specific error the idea is to create a .proto message like:

```
message AlreadyAtLatestVersionError {}
```

and put the `AlreadyAtLatestVersionError` message into the `Details` field of the `status.Status` to tag the error

```go
	if err := upgradePlatform(pm, ref, downloadCB, taskCB, req.GetSkipPostInstall()); err != nil {
		if err == ErrAlreadyLatest {
			status, e := status.Convert(err).WithDetails(&rpc.AlreadyAtLatestVersionError{})
			if e != nil { // should never happen
				panic(e)
			}
			return nil, status
		}
		return nil, status.Convert(err)
	}
```

on the client side we should do a type assertion to verify if the error is an `AlreadyAtLatestVersionError`, so something like:

```go
		if d := err.Details(); len(d) > 0 && d[0].(*rpc.AlreadyAtLatestVersionError) != nil {
			feedback.Printf("Platform %s is already at the latest version", platformRef)
		} else if err != nil {
			feedback.Errorf("Error during upgrade: %v", err)
			os.Exit(errorcodes.ErrGeneric)
		}
```

probably in this case it may be worth it to change the function `upgradePlatform` to return directly a `status.Status`, so we can avoid passing through the intermediate `ErrAlreadyLatest`, it will be something like:

```go
	if !latest.Version.GreaterThan(installed.Version) {
		status, e := status.New(codes.AlreadyExists, "platform already at latest version").WithDetails(&rpc.AlreadyAtLatestVersionError{})
		if e != nil { // should never happen
			panic(e)
		}
		return status
	}
```

or even:

```go
	if !latest.Version.GreaterThan(installed.Version) {
		status, _ := status.New(codes.AlreadyExists, "platform already at latest version").WithDetails(&rpc.AlreadyAtLatestVersionError{})
		return status
	}
```

since the `WithDetails` will never fail in this case.

Reply:
Thanks so much for your kind guidance on this @cmaglie . I have now added it to the PR just as you suggested: https://github.com/cmaglie/arduino-cli/pull/14/commits/410406fa43019ea05d7fa61dbaafec350ad8d679

There was one thing that didn't seem right to me though. The type assertion would cause a panic if the detail type was other than the expected one. It seems the better behavior in this case would be to treat it as a generic error, just as is done for any other non-`nil` status. So I made that change in a separate commit: https://github.com/cmaglie/arduino-cli/pull/14/commits/fcc316a90dff1ecae7020b2fea52e867cd866f33